ramblings/scenario_variables_etc.md

```diff
@@ -0,0 +1,94 @@
+# Scenario / variables / other filters etc
+
+At the moment we have scenarios and years hard-coded as things that can change the output on the map. We want to generalise this.
+
+I'm struggling to think of a decent name for this ... manifesto, but I'm going formalise what we're doing when we're painting pretty colours on the map a little bit. Heckling is requested and desired.
+
+# What is a model / survey data?
+
+I'm going to be lazy and pretend that predictions and observations are the same (which philosophically makes sense to me - observations are a guess at ground truth). So I'll call them all models.
+
+What is a model? It maps a series of independent variables (things you can change) to dependent variables (things you measure or predict).
+
+```
+model_[dependent_variable](independent_variables...) = {zoneID => some number}
+```
+
+So, for example,
+
+```
+model_kgco2perKM(scenario="the-squirrels-fight-back", year=1862, transportMode="jetpack", passengerAges="18-27") = {... => 9001}
+```
+
+Our key assumption is that each model has consistent units (i.e. different sets of criteria are directly comparable for the same dependent variable). We also assume that addition makes sense for all outputs.
+
+# How does this map to CSVs?
+
+Each CSV column (aside from the origin and destination zone IDs) corresponds to one set of `dependent_variable-independent_variables...`. Using the example from above:
+
+```
+model_kgco2perKM(scenario="the-squirrels-fight-back", year=1862, transportMode="jetpack", passengerAges="18-27") = {... => 9001}
+
+# corresponds to the column name
+kgco2perKM_the-squirrels-fight-back_1862_jetpack_18-27
+```
+
+Open question: how do we deal with missing independent variables? i.e. holes in the domain of the model? E.g. `scenario=millenium_bug` may not have `passengerAges`.
+
+Suggestion: if we want to support this, use a simple `key=value` format for column names, e.g. `kgco2perKM_scenario=the-squirrels-fight-back_transportMode=jetpack_...`.
```

Comment:
This is a question I think we need to answer. The PCT dataset has a bunch of columns that are not related to scenarios, but instead give context (e.g. hilliness of the best route between these zones), and I think that will be common.

We need to come up with:

- a way for data-uploaders to specify that to us
- a format to store that in and manipulate it
- decision on who implements the filtering logic (I think the simplest is that the frontend does all the filtering and just passes back a column id to the server)
- a way to add help and descriptions to the different filters

Reply:
I think the format of `dependent-variable_[independent-variables...]` where there is always exactly one independent variable is a nice easy way of solving this.

I think the biggest question is whether dependent variables will always have all independent variables or none; if they sometimes have a few then I think we'll have to do the `key=value` naming convention.

Explicitly:

```
# 1): Column with all independent variables - like "positional arguments"
#   NB: some pairs of values of independent variables may be missing, e.g., say I have scenarios and years, then not all scenarios may be available for all years.
dependent-variable_independent-variable-1_independent-variable-2_independent-variable-3_..._independent-variable-N

# 2): Column with none (i.e. constants)
dependent-variable

# 3): Column with only some relevant inputs - like "keyword arguments"
dependent-variable_independent-variable-1=value-1_independent-variable-4=value-2
```

All could co-exist peacefully.

I think in terms of UI, we should display but grey out values that are available for the current / base being changed but not available for the base / current that is staying the same.

From a user's perspective it would be nice if you could click on these greyed out values; the whole map could then grey out and the invalid selections on the opposite side highlighted in e.g. red so that they can be changed. But that is probably a bit tricky to do well.

I think that answers your first two bullet points.

---

Filtering: I'm not sure on this. If we do it in the frontend we'll probably essentially want a kind of sparse tensor that just enumerates all permitted combinations of variables? So really just a list of them all? It depends on what size we're expecting. If like GCVT we have approx 10x5x10, that's fine. If it's more like 100^4 then we'll definitely want the sparse route.

It could be simpler to do it in the backend: when a dependent variable is selected on the frontend, just send back the first combination of variables that matches and use that. Then when a user hovers and clicks on the drop-down menu, ask the server what values of that specific variable are valid. Then we'll have a linear rather than polynomial problem.

---

A way to add help and descriptions to the filters: this is the same problem as pretty-printing the names. So something like metadata.yml (or a wizard-made version of that) will be fine. We could include KaTeX / MathJax in the map so people can do superscripts etc

E.g.:

```
outcomes:
    kgco2perkm:
        pretty-name: "kg CO2 per km"
        description: "this is bad"
        colours: "greenred"
        flipscale: false
    redsquirrelpopulation:
        pretty-name: "Red squirrels per $km^2$"
        colours: "redwhite"
        flipscale: true

inputs:
    scenarios:
        1-the-squirrels-fight-back:
            pretty-name: "The squirrels fight back"
            description: "Humans have been sterilising grey squirrels for generations with laced peanut butter. Only a hard core remain. They decide that enough is enough and launch a vicious counter-campaign across TV & radio led by Saatchi&Saatchi"
        2-the-squirrels-dont-like-peanut-butter:
            pretty-name: "The squirrels don't like peanut butter"
            description: "The humans launch a sterilisation campaign against grey squirrels through laced peanut butter. It emerges that the squirrels don't much like peanut butter and nothing much happens."
        3-chaos-with-ed-miliband:
            pretty-name: "Chaos with Ed Miliband"
            description: ""
    years:
        1997:
            description: "Things can only get better"
        2020:
            description: "Things can't get worse"
```

If we want specific combinations of variables to have specific descriptions ... that would be harder. I think perhaps so difficult that we shouldn't entertain it unless we're asked.

---

I don't think we need the uploader to specify explicitly which combinations of things are permitted. We should be able to work that out automatically.